helm/cluster/templates/_helpers.tpl

```diff
@@ -151,4 +151,115 @@ Where `data` is the data to hash and `global` is the top level scope.
+{{- else if ($.Values.cluster).internal }}
+  {{- $internalValues = $.Values.cluster.internal }}
+{{- end }}
+{{- /* Create $.GiantSwarm.internal object */}}
+{{- if not $.GiantSwarm }}
+  {{- $_ := set $ "GiantSwarm" dict }}
+{{- end }}
+{{- /* Finally set "UseReleases" property that can be used in other templates */}}
+{{- $_ := set $.GiantSwarm "internal" $internalValues }}
+{{- end }}
+
+{{- define "cluster.internal.get-release-resource" }}
+{{- $_ := include "cluster.internal.get-internal-values" $ }}
+{{- $renderWithoutReleaseResource := ((($.GiantSwarm.internal).ephemeralConfiguration).offlineTesting).renderWithoutReleaseResource | default false }}
+{{- $releaseVersion := $.Values.global.release.version | trimPrefix "v" }}
+{{- $releaseVersion = printf "v%s" $releaseVersion }}
+{{- $release := lookup "release.giantswarm.io/v1alpha1" "Release" "" $releaseVersion }}
+{{- if $release }}
+  {{- $_ := set $.GiantSwarm "Release" $release }}
+{{ else if not $renderWithoutReleaseResource }}
+  {{- fail (printf "Release resource '%s' not found" $releaseVersion) }}
+{{- end }}
+{{- end }}
+
+{{- define "cluster.app.version" }}
+{{- $appVersion := "N/A" }}
+{{- $_ := (include "cluster.internal.get-release-resource" $) }}
+{{- if $.GiantSwarm.Release }}
+{{- range $_, $app := $.GiantSwarm.Release.spec.apps }}
+{{- if eq $app.name $.appName }}
+{{- $appVersion = $app.version }}
+{{- end }}
+{{- end }}
+{{- end }}
+{{- $appVersion }}
+{{- end }}
+
+{{- define "cluster.component.version" }}
+{{- $componentVersion := "N/A" }}
+{{- $_ := (include "cluster.internal.get-release-resource" $) }}
+{{- if $.GiantSwarm.Release }}
+{{- range $_, $component := $.GiantSwarm.Release.spec.components }}
+{{- if eq $component.name $.componentName }}
+{{- $componentVersion = $component.version }}
+{{- end }}
+{{- end }}
+{{- end }}
+{{- $componentVersion }}
+{{- end }}
+
+{{- define "cluster.component.kubernetes.version" }}
+{{- $_ := include "cluster.internal.get-provider-integration-values" $ }}
+{{- if $.GiantSwarm.providerIntegration.useReleases }}
+{{- $_ := set $ "componentName" "kubernetes" }}
+{{- include "cluster.component.version" $ | trimPrefix "v" }}
+{{- else if $.GiantSwarm.providerIntegration.kubernetesVersion }}
+{{- $.GiantSwarm.providerIntegration.kubernetesVersion | trimPrefix "v" }}
+{{- else }}
+{{- fail "Cannot determine Kubernetes version" }}
+{{- end }}
+{{- end }}
+
+{{- define "cluster.component.flatcar.version" }}
+{{- $_ := include "cluster.internal.get-provider-integration-values" $ }}
+{{- if $.GiantSwarm.providerIntegration.useReleases }}
+{{- $_ := set $ "componentName" "flatcar" }}
+{{- $flatcarVersion := include "cluster.component.version" $ | trimPrefix "v" }}
+{{- $flatcarVersion }}
+{{- else if $.GiantSwarm.providerIntegration.osImage }}
+{{- $.GiantSwarm.providerIntegration.osImage.version }}
+{{- else }}
+{{- fail "Cannot determine Flatcar version" }}
+{{- end }}
+{{- end }}
+
+{{- define "cluster.component.flatcar.variant" }}
```

Comment:
Similar like for `cluster.component.flatcar.version`.

One caveat here, since `Release.spec.components[].version` must be a valid semver, and we use plain numbers for variants (like `1`, `2`, `3`), we would put something like `1.0.0`, `2.0.0`, `3.0.0`, and then just use major version number here.

Or alternatively, maybe better, we actually really use regular sember version number for variant number.

P.S. Variant (or the number in the image name suffix) is something that people just set manually when they build a custom image (by default the variant number is not set), so we can simply decide how it must look like when somebody does this.